token/src/state.rs

```diff
@@ -17,38 +17,29 @@ use std::mem::size_of;
 #[repr(C)]
 #[derive(Clone, Copy, Debug, Default, PartialEq)]
 pub struct Token {
-    /// The total supply of tokens.
+    /// Token details.
     pub info: TokenInfo,
     /// Optional owner, used to mint new tokens.  The owner may only
     /// be provided during mint creation.  If no owner is present then the mint
     /// has a fixed supply and no further tokens may be minted.
     pub owner: COption<Pubkey>,
 }
 
-/// Delegation details.
-#[repr(C)]
-#[derive(Clone, Copy, Debug, Default, PartialEq)]
-pub struct AccountDelegate {
-    /// The source account for the tokens.
-    pub source: Pubkey,
-    /// The original maximum amount that this delegate account was authorized to spend.
-    pub original_amount: u64,
-}
-
 /// Account that holds tokens or may delegate tokens.
 #[repr(C)]
 #[derive(Clone, Copy, Debug, Default, PartialEq)]
 pub struct Account {
-    /// The type of token this account holds.
-    pub token: Pubkey,
-    /// Owner of this account.
+    /// The mint associated with this account
+    pub mint: Pubkey,
+    /// The owner of this account.
     pub owner: Pubkey,
-    /// Amount of tokens this account holds.
+    /// The amount of tokens this account holds.
     pub amount: u64,
-    /// If `delegate`  is None, `amount` belongs to this account.
-    /// If `delegate` is Option<_>, `amount` represents the remaining allowance
-    /// of tokens this delegate is authorized to transfer from the `source` account.
-    pub delegate: COption<AccountDelegate>,
+    /// If `delegate` is `Some` then `delegated_amount` represents
+    /// the amount authorized by the delegate
+    pub delegate: COption<Pubkey>,
+    /// The amount delegated
+    pub delegated_amount: u64,
```

Comment:
wdyt about wrapping the amount up in the `COption` as well?
```suggestion
    /// If `delegate` is `Some` then `delegated_amount` represents
    /// the amount authorized by the delegate and the amount delegated
    pub delegate: COption<(Pubkey, u64)>,
```

Reply:
I would be more inclined to do something like
```rust
put struct DelegateInfo {
    pub key: Pubkey,
    pub amount: u64,
};
pub struct Account {
...
    pub delegate: COption<DelegateInfo>
}
```

Reply:
Yeah that's definitely better than an unnamed tuple